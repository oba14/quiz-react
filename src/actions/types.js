export const GET_QUESTIONS = 'GET_QUESTIONS';
export const ERROR_FETCHING = 'ERROR_FETCHING';
export const IS_FETCHING = 'IS_FETCHING';
export const SET_CURRENT_ANSWER = 'SET_CURRENT_ANSWER';
export const SET_CURRENT_QUESTION = 'SET_CURRENT_QUESTION';
export const SET_ERROR = 'SET_ERROR';
export const SET_SHOW_RESULTS = 'SET_SHOW_RESULTS';
export const SET_ANSWERS = 'SET_ANSWERS';
export const RESET_QUIZ = 'RESET_QUIZ';