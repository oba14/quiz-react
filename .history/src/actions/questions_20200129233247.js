import axios from 'axios';
const url = "https://opentdb.com/api.php?amount=10&category=11&difficulty=hard&type=multiple"

// Fetch ALL FORMS from mongodb and display item when component is rendered
export const getQuestions = () => {
    return async (dispatch) => {
        dispatch({
            type: 'IS_FETCHING'
        })
        
        await axios(`${url}`)
          .then(res => {
            dispatch({
                type: 'GET_QUESTIONS',
                payload: res.data,
            })
            }) 
          .catch(error => { 
            console.log("Canâ€™t access " + url + " response. Blocked by browser?")
            dispatch({
                type: 'ERROR_FETCHING',
                payload: error.message
            })
        })
    }
}
