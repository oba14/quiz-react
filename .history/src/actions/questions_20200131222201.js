import axios from 'axios';
import {GET_QUESTIONS, 
    ERROR_FETCHING, 
    IS_FETCHING, 
    SET_ERROR, 
    SET_ANSWERS, 
    SET_CURRENT_ANSWER,
    SET_CURRENT_QUESTION,
    SET_SHOW_RESULTS
} from './types'

//const url = "https://opentdb.com/api.php?amount=10&category=11&difficulty=hard&type=multiple&encode=url3986"

// Fetch ALL FORMS from mongodb and display item when component is rendered
export const getQuestions = (data) => {
    return async (dispatch) => {
        dispatch({
            type: IS_FETCHING
        })
        
        await axios(`https://opentdb.com/api.php?amount=${data.noOfQuestions}&category=${data.selectedCategory}&difficulty=${data.selectedDifficulty}&type=multiple&encode=url3986`)
          .then(res => {
            dispatch({
                type: GET_QUESTIONS,
                payload: res.data.results,
            })
            }) 
          .catch(error => { 
            console.log("Canâ€™t access " + url + " response. Blocked by browser?")
            dispatch({
                type: ERROR_FETCHING,
                payload: error.message
            })
        })
    }
}

export const setError = () => dispatch => {
    dispatch({type: SET_ERROR,
        error: 'Please select an option'
    });
}

export const setAnswers = (answers) => dispatch => {
    dispatch({
        type: SET_ANSWERS, 
        answers
    })
}

export const setCurrentAnswer = (e) => dispatch => {
    dispatch({
        type: SET_CURRENT_ANSWER, 
        currentAnswer: e
    })
}

export const setCurrentQuestion = (currentQuestion) => dispatch => {
    dispatch({
        type: SET_CURRENT_QUESTION,
        currentQuestion: currentQuestion + 1,
    })
}

export const setShowResults = () => dispatch => {
    dispatch({
        type: SET_SHOW_RESULTS, 
        showResults: true
    })
}