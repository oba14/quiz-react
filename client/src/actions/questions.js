import axios from "axios";
import {
  GET_QUESTIONS,
  ERROR_FETCHING,
  IS_FETCHING,
  SET_ERROR,
  SET_ANSWERS,
  SET_CURRENT_ANSWER,
  SET_CURRENT_QUESTION,
  SET_SHOW_RESULTS,
  QUIT,
  RESET_QUIZ
} from "./types";

const url = "http://localhost:5000/quiz";

export const questions = data => {
  return {
    type: GET_QUESTIONS,
    payload: data
  };
};

export const isfetching = () => {
  return {
    type: IS_FETCHING
  };
};

export const errorFetching = err => {
  return {
    type: ERROR_FETCHING,
    payload: err
  };
};

export const getQuestions = data => {
  console.log("GET QUESTION ACTION BEING CALLED", data);

  return async dispatch => {
    dispatch(isfetching());
    try {
      return await axios({
        method: "post",
        url: url,
        data: {
          noOfQuestions: data.noOfQuestions,
          selectedCategory: data.selectedCategory,
          selectedDifficulty: data.selectedDifficulty
        }
      })
        .then(res => {
          console.log("INSIDE THENNNNN", res);
          dispatch(questions(res.data.results));
        })
        .catch(error => {
          console.log("Canâ€™t access " + url + " response. Blocked by browser?");
          dispatch(errorFetching(error.message));
        });
    } catch (error) {
      dispatch(errorFetching(`catch block error: ${error.message}`));
    }
  };
};

// export const setError = () => dispatch => {
//   dispatch({ type: SET_ERROR, error: 'Please select an option' });
// };

export const setError = () => {
  return { type: SET_ERROR, error: "Please select an option" };
};

export const setAnswers = answers => dispatch => {
  dispatch({
    type: SET_ANSWERS,
    answers
  });
};

export const setCurrentAnswer = e => dispatch => {
  dispatch({
    type: SET_CURRENT_ANSWER,
    currentAnswer: e
  });
};

export const setCurrentQuestion = currentQuestion => dispatch => {
  dispatch({
    type: SET_CURRENT_QUESTION,
    currentQuestion: currentQuestion + 1
  });
};

export const setShowResults = () => dispatch => {
  dispatch({
    type: SET_SHOW_RESULTS,
    showResults: true
  });
};

export const quitQuiz = () => dispatch => {
  dispatch({
    type: QUIT
  });
};

export const resetQuiz = () => dispatch => {
  dispatch({
    type: RESET_QUIZ
  });
};
