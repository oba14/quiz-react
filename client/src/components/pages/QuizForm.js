import React, { useEffect } from "react";
import { useDispatch } from "react-redux";
import { useForm } from "react-hook-form";
import { getQuestions } from "../../actions/questions";

const QuizForm = props => {
  const dispatch = useDispatch();
  const { register, handleSubmit, errors } = useForm();

  const categories = [
    { id: 9, name: "General Knowledge" },
    { id: 10, name: "Entertainment: Books" },
    { id: 11, name: "Entertainment: Film" },
    { id: 12, name: "Entertainment: Music" },
    { id: 13, name: "Entertainment: Musicals & Theatres" },
    { id: 14, name: "Entertainment: Television" },
    { id: 15, name: "Entertainment: Video Games" },
    { id: 16, name: "Entertainment: Board Games" },
    { id: 17, name: "Science & Nature" },
    { id: 18, name: "Science: Computers" },
    { id: 19, name: "Science: Mathematics" },
    { id: 20, name: "Mythology" },
    { id: 21, name: "Sports" },
    { id: 22, name: "Geography" },
    { id: 23, name: "History" },
    { id: 24, name: "Politics" },
    { id: 25, name: "Art" },
    { id: 26, name: "Celebrities" },
    { id: 27, name: "Animals" },
    { id: 28, name: "Vehicles" },
    { id: 29, name: "Entertainment: Comics" },
    { id: 30, name: "Science: Gadgets" },
    { id: 31, name: "Entertainment: Japanese Anime & Manga" },
    { id: 32, name: "Entertainment: Cartoon & Animations" }
  ];

  const addOptions = () => {
    const categoryDom = document.getElementById("categories");
    categories.map(category => {
      categoryDom.innerHTML += `<option value=${category.id}>${category.name}</option>`;
    });
  };

  const onSubmit = () => {
    const myForm = document.getElementById("myForm");
    const formData = new FormData(myForm);
    const selectedCategory = formData.get("category");
    const selectedDifficulty = formData.get("difficulty");
    const noOfQuestions = formData.get("noofquestions");

    // get all form entries
    // const data = formData.entries();
    // console.log("FORM ENTRIES", data.next());

    // var obj = data.next();
    // var retrieved = {};
    // while (undefined !== obj.value) {
    //   retrieved[obj.value[0]] = obj.value[1];
    //   obj = data.next();
    // }
    // console.log("retrieved: ", retrieved);

    dispatch(
      getQuestions({
        selectedCategory,
        selectedDifficulty,
        noOfQuestions
      })
    );

    props.history.push("/quiz");
  };

  useEffect(() => {
    addOptions();
  });

  return (
    <div className="container">
      <h3>Select quiz category and difficult level</h3>
      <form
        onSubmit={handleSubmit(onSubmit)}
        id={"myForm"}
        name={"myForm"}
        encType="multipart/form-data"
      >
        <label>1- Select a category</label> <br></br>
        <select
          style={{ width: "47%" }}
          id={"categories"}
          name="category"
          ref={register({ required: true })}
        ></select>{" "}
        <br></br>
        <label>2- Select a difficulty level</label> <br></br>
        <select
          style={{ width: "47%" }}
          name="difficulty"
          ref={register({ required: true })}
        >
          <option value="easy">Easy</option>
          <option value="medium">Medium</option>
          <option value="hard">Hard</option>
        </select>
        <br></br>
        <label>3- Select number of questions.</label> <br></br>
        <input
          style={{ width: "47%", margin: "0 auto" }}
          className="form-control"
          type="number"
          placeholder="Number of questions"
          name="noofquestions"
          ref={register({ required: true })}
          required
        />
        <br />
        <button className="btn btn-primary" type="submit">
          {" "}
          Submit{" "}
        </button>
      </form>
    </div>
  );
};
export default QuizForm;
