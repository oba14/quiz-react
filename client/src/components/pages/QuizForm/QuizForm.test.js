import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import QuizForm from './QuizForm';
import { getQuestions } from '../../../actions/questions';

const mockStore = configureStore([]);
describe('My Connected React-Redux Component', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      questions: [],
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <QuizForm />
      </Provider>
    );
  });
  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should dispatch an action on button click', () => {
    renderer.act(() => {
        component.root.findByType('button').props.onClick();
      });

      renderer.act(() => {
        component.root.findByType('input')
          .props.onChange({ target: { value: 2 } });
      });
      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(
        getQuestions({ payload: 'sample text' })
      );
  });
});