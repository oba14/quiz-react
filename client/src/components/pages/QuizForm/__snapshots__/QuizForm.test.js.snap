// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="container"
>
  <h3
    data-testid="quiz-form-heading"
  >
    Select quiz category and difficult level
  </h3>
  <form
    encType="multipart/form-data"
    id="myForm"
    name="myForm"
    onSubmit={[Function]}
  >
    <label>
      1- Select a category
    </label>
     
    <br />
    <select
      id="categories"
      name="category"
      style={
        Object {
          "width": "47%",
        }
      }
    />
     
    <br />
    <label>
      2- Select a difficulty level
    </label>
     
    <br />
    <select
      name="difficulty"
      style={
        Object {
          "width": "47%",
        }
      }
    >
      <option
        value="easy"
      >
        Easy
      </option>
      <option
        value="medium"
      >
        Medium
      </option>
      <option
        value="hard"
      >
        Hard
      </option>
    </select>
    <br />
    <label>
      3- Select number of questions.
    </label>
     
    <br />
    <input
      className="form-control"
      name="noofquestions"
      placeholder="Number of questions"
      required={true}
      style={
        Object {
          "margin": "0 auto",
          "width": "47%",
        }
      }
      type="number"
    />
    <br />
    <button
      className="btn btn-primary"
      data-testid="form-submit-btn"
      type="submit"
    >
       
      Submit
       
    </button>
  </form>
</div>
`;
